[LW-9145] TxSummary refactor

VanessaPC opened the pull request:
# Checklist

- [x] JIRA - LW-9145 LW-9710
- [ ] Proper tests implemented
- [x] Screenshots added.

---

## Proposed solution
- Refactor the transaction summary for dapps 
- add to and from sections 
- txType and origin 

## Testing
- Run the app 
- Go to wingriders and try to get some wingrider token from ada, there shouldn't be tokens 
- Do the reverse transactions, you should see the token 
- Go to handle.me , mint a token, you should see the transaction 


**Tx notes:** 
- A transaction doesn't have to have to and from addresses. Sometimes it may have one sections with various, and sometimes both. But not always. 
- Transactions from swapping places such as Wingrider or SundaeSwap work with a batcher which means that you might see an amount in their interface, and a slightly different amount of ADA or tADA in the dapp, because you will see the amount of coins that don't get sent to a batcher to process the following transaction to actually create the swap. 


## Screenshots

![Screenshot 2024-02-21 at 08 28 02](https://github.com/input-output-hk/lace/assets/4052063/e965b609-dcb7-4d2e-a92f-4546d70f37dd)
![Screenshot 2024-02-21 at 08 29 17](https://github.com/input-output-hk/lace/assets/4052063/7a3e440e-ef53-447e-8f9d-8885a914ef1d)
![Screenshot 2024-02-21 at 08 33 55](https://github.com/input-output-hk/lace/assets/4052063/d24ea5a3-5bfb-496a-a4ef-52e1c53fa2e3)

// This image is here only to show the display of to address / from address sections, the styles apply from the previous snapshots. 
<img width="400" alt="Screenshot 2024-02-14 at 08 35 05" src="https://github.com/input-output-hk/lace/assets/4052063/81b69b8a-fa43-4349-a89e-b0e069788b07">

DominikGuzei commented:
@VanessaPC it seems like you need to update this PR to the latest `main` as the changed file count jumped from around 50 to over 200 and i can see many unrelated changes on Github 🤔

pczeglik-iohk commented on packages/core/src/ui/components/DappTransaction/DappTransaction.tsx:
```diff
@@ -1,91 +1,177 @@
+    ? assetWithAmount.assetInfo.nftMetadata.name
+    : assetWithAmount.assetInfo.tokenMetadata.name;
+
+const hash = (assetWithAmount: AssetInfoWithAmount) =>
```
This is only suggestion from my side. Maybe we can rename `hash` and `tokenName` to `getHash` and `getTokenName` so we are consistent with `getStringFromLovelace` and `getTxType`.

What do you think?

DominikGuzei replied:
and maybe make it more descriptive like `getAssetHash`?

pczeglik-iohk commented on packages/core/src/ui/components/DappInfo/DappInfo.module.scss:
```diff
@@ -49,4 +49,4 @@
     }
   }
-}
+}
```
Can you revert this change please? :)

pczeglik-iohk commented on packages/core/src/ui/components/ActivityDetail/TransactionFooterDetails.tsx:
```diff
@@ -0,0 +1,62 @@
+/* eslint-disable no-magic-numbers */
```
where is this magic number?

pczeglik-iohk commented on packages/common/src/ui/lib/add-ellipsis.ts:
```diff
@@ -1,6 +1,13 @@
 /* eslint-disable no-magic-numbers */
 const addedLength = 4;
 
+export const truncate = (text: string, partOnelength: number, partTwoLength: number): string => {
```
`partOneLength`

pczeglik-iohk commented on apps/browser-extension-wallet/src/views/browser-view/features/staking/components/StakePoolDetails/StakePoolDetailsDrawer.tsx:
```diff
@@ -39,7 +39,6 @@ export const StakePoolDetailsDrawer = ({
   const { setIsRestaking } = useSubmitingState();
   const { isInMemoryWallet } = useWalletStore();
   const { password, removePassword } = usePassword();
-
```
Can you revert this change, please :)

pczeglik-iohk commented on apps/browser-extension-wallet/src/utils/combined-input-resolvers.ts:
```diff
@@ -0,0 +1,15 @@
+  chainHistoryProvider?: any;
+}
+
+export const combinedInputResolver = ({
```
`combine`? if you like this suggestion, than we need to rename the file as well :)

DominikGuzei replied:
@pczeglik-iohk there is already a `combineInputResolvers` which this function wraps in order to provide a fixed, already combined set of resolvers, so i think the past tense makes sense since you cannot "combine" arbitrary input resolvers with it.

Maybe instead of the generic `combined` part we could say a bit more about the combination in the name? (i don't have any good suggestion atm)

pczeglik-iohk replied:
yes there is, but this one created by @VanessaPC does not contain `s` on the end :)

pczeglik-iohk commented on apps/browser-extension-wallet/src/hooks/useChainHistoryProvider.ts:
```diff
@@ -0,0 +1,15 @@
+import { chainHistoryHttpProvider } from '@cardano-sdk/cardano-services-client';
+import { logger } from '@lib/wallet-api-ui';
+
+type useChainHistoryProviderArgs = {
```
Should type start from the upper case?

As this type is local, maybe it's ok to have `ChainHistoryProviderArgs`?

DominikGuzei replied:
or maybe even inlining the type into the function definition in this case?

pczeglik-iohk commented on apps/browser-extension-wallet/src/features/dapp/components/ConfirmTransaction.tsx:
```diff
@@ -86,52 +47,37 @@ export const ConfirmTransaction = withAddressBookContext((): React.ReactElement
-    }
-    return [...uniqueAssetIds.values()];
-  }, [req]);
+  const newChainHistoryProvider = useChainHistoryProvider({ chainName });
```
Why new chain history provider?

pczeglik-iohk commented on apps/browser-extension-wallet/src/features/dapp/components/ConfirmTransaction.tsx:
```diff
@@ -295,14 +204,16 @@ export const ConfirmTransaction = withAddressBookContext((): React.ReactElement
   return (
-    <Layout pageClassname={styles.spaceBetween} title={t(sectionTitle[DAPP_VIEWS.CONFIRM_TX])}>
-      {req && txSummary ? (
+    <div data-testid="layout-title" className={styles.transactionContainer}>
```
What do you think about using `FlexComponent` from ui toolkit, instead of div with custom defined styles?

pczeglik-iohk replied:
https://github.com/input-output-hk/lace/blob/main/packages/ui/src/design-system/flex/flex.component.tsx

pczeglik-iohk commented on packages/ui/src/design-tokens/theme/hooks/use-theme-variant.tsx:
```diff
@@ -0,0 +1,11 @@
+import type { ThemeColorScheme } from '../theme.context';
+
+export const useThemeVariant = (): { theme: ThemeColorScheme } => {
+  const themeContext = useContext(ThemeContext);
```
should we throw if no `themeContext`? like for example in `useAddressBookContext`

VanessaPC replied:
Here isn't needed because we have a default set.

pczeglik-iohk commented on packages/ui/src/design-system/grid/grid.component.tsx:
```diff
@@ -9,13 +9,15 @@ export type Props = PropsWithChildren<{
   columns = '$none',
   children,
   rows = '$none',
+  alignItems,
```
does it make sense to provide default value for `alignItems`?

VanessaPC replied:
I wouldn't because the Grid is used in other places without using this property, and this property is also not always used in a grid, if I set a default now, I will be causing issues with the places where its used.

pczeglik-iohk reviewed (commented):
I reviewed the code and posted a few comments.

Nothing crucial, but there are a few improvements which can lead to increased consistency and maintenance of the code.

Let me know what do you think :)

and thank you @VanessaPC for great work.

DominikGuzei commented on packages/core/src/ui/components/DappAddressSections/DappAddressSections.tsx:
```diff
@@ -0,0 +1,147 @@
+const charBeforeEllipsisName = 8;
+const charAfterEllipsisName = 8;
+
+export const DappAddressSections = ({
```
is this really only `DApp` related or could this address representation potentially be re-used in other scenarios?

VanessaPC replied:
I don't think this part could be reused, the UI component to display assets could. I think that the TX history does something different, or does nothing at the moment. If we re-do that part and need to use this, then we can think about abstracting this further.

DominikGuzei commented on packages/ui/src/design-system/grid/grid.css.ts:
```diff
@@ -83,6 +83,11 @@ export const grid = recipe({
     },
+    alignItems: {
+      $center: {
+        alignItems: 'center',
```
does this mean we can only center the grid for now?

VanessaPC replied:
I added the basic properties now.

VanessaPC commented on packages/icons/src/AdaComponent.tsx:
```diff
@@ -0,0 +1,28 @@
+/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
+import * as React from 'react';
+import type { SVGProps } from 'react';
+const SvgAdacomponent = (props: SVGProps<SVGSVGElement>) => (
```
is the convention used in this folder, `SvgLaceLogoComponent`, `SvgUploadComponent`...

pczeglik-iohk replied:
sorry, I wasn't clear enough

VanessaPC replied:
oh!... pushed now.

VanessaPC commented:
@DominikGuzei @pczeglik-iohk this is ready for re-review.

pczeglik-iohk reviewed (approved):
Code looks great. Thank you @VanessaPC

DominikGuzei commented on apps/browser-extension-wallet/src/hooks/useChainHistoryProvider.ts:
```diff
@@ -0,0 +1,15 @@
+import { chainHistoryHttpProvider } from '@cardano-sdk/cardano-services-client';
+import { logger } from '@lib/wallet-api-ui';
+
+type UseChainHistoryProviderArgs = {
```
I can see that the names of the networks are all over the place in our code base, maybe let's move them into an enum somewhere like it's already one for the e2e here https://github.com/input-output-hk/lace/blob/6516722024494584fd3bb81196a345d5895f43db/packages/e2e-tests/src/types/network.ts

VanessaPC replied:
Your suggestions is really good. This has come up in conversation already somewhere else, but is there a generic constants place for an enum of this type? and have it been considered a constants file/folder in the app?

DominikGuzei replied:
@VanessaPC I'm not sure, it's definitely something we could discuss with the team 👍

VanessaPC replied:
I have created: https://input-output.atlassian.net/browse/LW-9883 to address the topic in a general setting. I think that doing a cleanup like this shouldn't be too complicated and it'd be better done at once.

DominikGuzei reviewed (approved):
Works great now 🎉 I just left one small suggestion that you can optionally look into (not really the scope of this PR)

VanessaPC committed:
feat: transaction representation refactor
- transaction includes transaction fees
- it displays returned deposit and deposits
- it shows from address sections wiht tokens and nfts
- it shows to address section with tokens and nfts
- it displays tokens and nfts in the summary
- it summarises the balance in coins of a transactions
- it displays the dapp name in the origin tab

pczeglik-iohk commented on apps/browser-extension-wallet/src/features/dapp/components/confirm-transaction/DappTransactionContainer.tsx:
```diff
@@ -104,10 +104,7 @@ export const DappTransactionContainer = withAddressBookContext(
-      () => userRewardAccounts && userRewardAccounts.map((key) => key.address),
-      [userRewardAccounts]
-    );
+    const rewardAccountsAddresses = useMemo(() => userRewardAccounts?.map((key) => key.address), [userRewardAccounts]);
```
thank you @VanessaPC

VanessaPC committed:
Merge branch 'main' into feat/lw-9145-tx-summary-refactor
